Adding autonight detection using software (to be tested)

chisht opened the pull request:
Copy /config/autonight.conf.sw to /config/autonight.conf to enable software
version of autonight detection

chisht commented:
Added a new commandline to existing autonight. If -S is specified, it'll switch to software autonight. Default is existing one.

Please test.

jmtatsch commented:
Nice! Could you also make a PR with the changed sources?

chisht commented:
@jmtatsch PR for source changes is up.

badabing2005 commented:
I reverted to the older version that you posted as I find this latest one constantly switching when it is semi dark, the switching is quite rapid (every second or so, sometimes even more), probably because as it turns the ir led on/off or flips the ir cut the average measured value is changing which is causing it to switch back. (just a guess)

Considering that the three switches, toggle night mode, toggle IR Led, toggle IR cut are not done in parallel (all at once) which is understandable,
Perhaps there is an optimal order to avoid constant on / off

Just a guess here.
If light level is high (IR cut on, IR Led off, night off) and we need to go to dark mode (IR cut off, IR led on, night on)
We go in this order:
- Night on, to turn it to black and  white, which avoids the average level to be changed much, which also helps to avoid getting pink images.
- IR cut off (which will only increase the average level if there is other IR light source
- IR Led On

Night to day would be the exact reverse of the above
Turn off IR Led, then turn on IR cut and only then turn the Night off.

Thanks

bachoo786 commented:
how do I add /config/autonight.conf.sw to /config/autonight.conf to enable software
version of autonight detection?

chisht commented:
@bachoo786 , yes it is working for me. As for config file, copy /config/autonight.conf.sw to /config/autonight.conf on your sdcard.

chisht commented:
@badabing2005 , Thanks for testing and the comment. The only difference between the old and new version is old had a bug in reading the ISP info before the driver is ready to return the correct data. So I had a 10 sec delay right in the start only. As far as I remember, I did not change anything else.

As for switching back and forth every second. Code actually does wait for exposure to settle down before switching to day mode. But one thing that'd make it switch right away is the White Balance Color temperature. If that is higher than 3000, it'd switch back to day mode. Because that low color temperature can not be caused by IR only. I'm guessing that logic needs some improving.

One thing I'd try is to make all those configuration parameters (including the WB color temp) as commandline options. So you can tweak them according to your lighting. BTW you can also see what's causing it to switch mode by running in manually from the shell as "autonight -S -v" . And more v to increase the verbosity.

Good idea about using three modes. I'd need to think about it.

BTW if you understand C, feel free to take a look at the code and hopefully improve.

bachoo786 commented:
@chisht sorry for my novelty but are you saying to remove my SD card and copy the autonight.conf.sw to autonight.conf? and where are these files that I need to copy?

thanks.

chisht commented:
@bachoo786 , that file is in this PR at https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/blob/18e9002aedcef9f3b627ccf368a8629fc3f57c3e/firmware_mod/config/autonight.conf.sw
Or you can just create one on your own. It only needs to have "-S " without quotes.

BTW you did get the new autonight binary from this PR, right?

bachoo786 commented:
@chisht nope I have not got anything from this PR. Do I have to get it? I was thinking the procedure was to just add the autonight.conf to config folder?

chisht commented:
@bachoo786 , click on the "files changed" tab at the top (alognwith conversation). That should show two files. Download them both. Copy autonight bin to the sdcard as /bin/autonight. And autonight.conf.sw to the sdcard as /config/autonight.conf

ZipperZ commented:
I'm trying it out. And I have a condition where it goes like crazy on and off... Some hysteresis should be implemented

chisht commented:
@ZipperZ , Likely whats getting in and out of night mode are two different variables. Hysteresis won't solve that.

bachoo786 commented:
@chist I am having the same issues as @ZipperZ and to debug your way is confusing..

badabing2005 commented:
Thanks @chisht for all the effort and the support.

I"ll be playing and testing this tomorrow, but I first wanted to quickly give you some of my output in case you notice something obvious.

With lights on, Camera was in day mode, and the I turned the lights off, there was some far background light, and also some light reflection from a monitor close to the camera.
And my night mode (as you can see in the logs) was constantly flipping.
I would get nice gray detailed image with IR on, and dark details showing, then it would flip to pink (ie no ir cut) or totally dark gray (almost black, ir led off) then one of the others, and even at times, I would get grainy colored image.

```
[root@Wyze:/system/sdcard/bin]# ./autonight -S -vv
Night Mode Disabled
Exposure falling by -19% to 1104405 count 0
Exposure falling by -13% to 1258155 count 0
Exposure 1258155, 17>
Night Mode Enabled
Exposure falling by 24% to 944254 count 0
Exposure changed by 0% to 946751,14 (944254) 16 3300 897041.300000
Exposure up to 946751 from 944254
Exposure changed by 0% to 949024,12 (946751) 16 3200 899413.450000
Exposure up to 949024 from 946751
Colortemp 3100
Night Mode Disabled
Exposure falling by -32% to 1254974 count 0
Exposure 1254974, 5>
Night Mode Enabled
Exposure falling by 22% to 967800 count 0
Exposure changed by 0% to 968277,10 (967800) 17 3500 919410.000000
Exposure up to 968277 from 967800
Exposure changed by 0% to 969348,8 (968277) 17 3500 919863.150000
Exposure up to 969348 from 968277
Exposure changed by 0% to 970061,8 (969348) 17 3400 920880.600000
Exposure up to 970061 from 969348
Exposure changed by 0% to 970299,7 (970061) 17 3300 921557.950000
Exposure up to 970299 from 970061
Exposure changed by 0% to 969125,8 (970299) 17 3300 921784.050000
Exposure changed by 0% to 968538,8 (970299) 17 3200 921784.050000
Colortemp 3100
Night Mode Disabled
Exposure falling by -30% to 1260265 count 0
Exposure 1260265, 5>
Night Mode Enabled
Exposure falling by 23% to 968173 count 0
Exposure changed by 0% to 967586,9 (968173) 15 3500 919764.350000
Exposure changed by 0% to 969013,8 (968173) 15 3400 919764.350000
Exposure up to 969013 from 968173
Exposure changed by 0% to 969370,8 (969013) 15 3400 920562.350000
Exposure up to 969370 from 969013
Exposure changed by 0% to 968312,8 (969370) 15 3300 920901.500000
Exposure changed by 0% to 967725,8 (969370) 15 3300 920901.500000
Exposure changed by 0% to 966786,8 (969370) 15 3200 920901.500000
Colortemp 3100
Night Mode Disabled
Exposure falling by -30% to 1262985 count 0
Exposure 1262985, 7>
Night Mode Enabled
Exposure falling by 23% to 969029 count 0
Exposure changed by 0% to 967972,9 (969029) 16 3500 920577.550000
Exposure changed by 0% to 969042,8 (969029) 16 3500 920577.550000
Exposure up to 969042 from 969029
Exposure changed by 0% to 969399,8 (969042) 16 3400 920589.900000
Exposure up to 969399 from 969042
Exposure changed by 0% to 968342,8 (969399) 16 3300 920929.050000
Exposure changed by 0% to 967403,8 (969399) 16 3200 920929.050000
Exposure changed by 0% to 967645,8 (969399) 16 3200 920929.050000
Colortemp 3100
Night Mode Disabled
Exposure falling by -30% to 1262985 count 0
Exposure 1262985, 8>
Night Mode Enabled
Exposure falling by 23% to 968452 count 0
Exposure changed by 0% to 961612,10 (968452) 16 3600 920029.400000
Exposure falling by 1% to 950845 count 0
Exposure changed by 0% to 950845,9 (968452) 16 3500 920029.400000
Exposure changed by 0% to 942889,9 (968452) 16 3300 920029.400000
```

Some questions:

> If in day mode and exposure is over 90%, switch to night mode.
What is considered 90% or higher in the above data?

> Iridix, From what I can gather from my experiments. This reports IR light coming back to the sensor. If IR is on and something is closer causing IR to reflect back, this number gets very high. But it usually also go slightly high, when normal room lights are on.

Iridix is not always displayed, see above data, in fact through my initial quick tests, I've only seen Iridix values reported on one occasion, when I had the lights in the room on.

In the above tests, I came close to the camera where my face was bright white due to IR reflection, and Iridix numbers were never reported.

**Color Temperature** is never reported when Night Mode is Disabled, it's only reported when it's enabled, how does once choose the right threshold without knowing the highs.

Ok -B is to set the Threshold for Color Temperature,
What about the other command line parameters
N
M
W
T
D
X

What do those represent?

Is it possible to tell the camera to ignore a region?
In one corner of the room, I have a TV, which  would like to ignore the light levels coming from the TV, I doubt that this is possible, but if it were, I'd get more stable lighting in the room with little variance

Off Topic, anyone knows if I can exclude the TV region from motion detection so that changes on the TV screen does not trigger screenshots (stills) 

Lastly, do you just call nightmode on nightmode off in common_functions.sh or do you control
IR LED
IR Cut
Grayscale (setconf)
in your binary?
If it is the latter? are they always triggered together? (in what order) or separately depending on the read data?
I'll check the source code if I'm asking too much.

Thanks
BB

chisht commented:
@badabing2005 , I'd reply later in detail to you questions. But you seem to be suffering from the same issue that @ZipperZ and @bachoo786 are. And I did manage to reproduce it at my end. Basically two things were going on. 

1. First if light in on in the next room, and some of it in sneaking into the camera view, color temperature drops to below 3000 even in night mode and it switches to day mode right away.  My solution is to note down the colortemp when we start night mode. And switch to day if it drops by -B parameter (default is 200).
2. But then I ran into second issue. Which was that after switching to night mode, I had one second delay for auto mode to settle down. But that was not enough.  In some of my tests, color temp was ~2900, 1 sec after enabling night mode. But then jumped to 3300 in two more seconds. So I increased 1sec delay to 3sec.

I updated the binary, hopefully this would fix the problem.

chisht commented:
BTW as for source code, feel free to check https://github.com/EliasKotlyar/Xiaomi-Dafang-Software/pull/33
Though I haven't updated it with the latest changes yet.

badabing2005 commented:
@chisht 

The latest binary is worse, much more frequent hysteria, and worst of all no output log, even with -vv
Sorry for the bad news, and thanks for your persistence to help us.

BB

chisht commented:
@badabing2005 I just ran it again. It sure does the same log as before.
You sure you running it from the shell as 
autonight -S -vv

badabing2005 commented:
> 
> 
> @badabing2005 I just ran it again. It sure does the same log as before.
> You sure you running it from the shell as
> autonight -S -vv

Sorry my bad, had forgotten the - in front of vv
a bit better, but still flipping

```[root@Wyze:/system/sdcard/bin]# autonight -S -vv
Night Mode Disabled
Exposure 1253711, 13>
Night Mode Enabled
Exposure falling by 22% to 973198 count 0
night mode start 973198, 8, 3100
Exposure falling by 1% to 962153 count 0
Exposure changed by 0% to 962153,8 (973198) 8 3100 924538.100000
Exposure changed by 0% to 965733,8 (973198) 8 3100 924538.100000
Exposure changed by 0% to 966565,8 (973198) 8 3100 924538.100000
Exposure changed by 0% to 967279,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 967279,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 967279,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 967397,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 963819,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 963305,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 964849,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 965324,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 965324,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 965800,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 966038,8 (973198) 8 3000 924538.100000
Exposure falling by 11% to 852047 count 0
Exposure changed by 0% to 852047,8 (973198) 8 3000 924538.100000
Exposure count 1 <973198 to 852047> Iridix <8 to 8>
Exposure falling by -3% to 884687 count 0
Exposure changed by 0% to 884687,8 (973198) 8 3100 924538.100000
Exposure count 1 <973198 to 884687> Iridix <8 to 8>
Exposure falling by -8% to 956835 count 0
Exposure changed by 0% to 956835,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 958742,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 958390,8 (973198) 8 3000 924538.100000
Exposure changed by 0% to 958273,9 (973198) 8 3000 924538.100000
Exposure falling by 1% to 943345 count 0
Exposure changed by 0% to 943345,9 (973198) 8 3000 924538.100000
Exposure changed by 0% to 942186,10 (973198) 8 3000 924538.100000
Exposure changed by 0% to 943866,10 (973198) 8 3000 924538.100000
Colortemp 2900
Night Mode Disabled
Exposure changed by 0% to 945197,10 (973198) 8 2900 924538.100000
Exposure falling by -32% to 1254534 count 0
Exposure 1254534, 6>
Night Mode Enabled
Exposure falling by 25% to 934072 count 0
night mode start 934072, 11, 3100
Exposure changed by 0% to 934666,11 (934072) 11 3100 887368.400000
Exposure up to 934666 from 934072
Exposure changed by 0% to 935142,11 (934666) 11 3100 887932.700000
Exposure up to 935142 from 934666
Exposure changed by 0% to 935499,10 (935142) 11 3100 888384.900000
Exposure up to 935499 from 935142
Exposure changed by 0% to 935499,10 (935499) 11 3000 888724.050000
Exposure changed by 0% to 928761,11 (935499) 11 3000 888724.050000
Exposure falling by 23% to 712459 count 0
Exposure changed by 0% to 712459,11 (935499) 11 3200 888724.050000
Exposure count 1 <935499 to 712459> Iridix <11 to 11>
Exposure falling by -30% to 928002 count 0
Exposure changed by 0% to 928002,11 (935499) 11 3300 888724.050000
Exposure changed by 0% to 931462,11 (935499) 11 3200 888724.050000
Exposure changed by 0% to 931580,11 (935499) 11 3200 888724.050000
Exposure changed by 0% to 930524,11 (935499) 11 3200 888724.050000
Exposure changed by 0% to 929115,11 (935499) 11 3100 888724.050000
Exposure changed by 0% to 929115,11 (935499) 11 3100 888724.050000
Exposure changed by 0% to 928648,11 (935499) 11 3000 888724.050000
Exposure changed by 0% to 928886,11 (935499) 11 3000 888724.050000
Colortemp 2900
Night Mode Disabled
Exposure changed by 0% to 928772,11 (935499) 11 2900 888724.050000
Exposure falling by -35% to 1255611 count 0
Exposure 1255611, 7>
Night Mode Enabled
Exposure falling by 25% to 930483 count 0
night mode start 930483, 11, 3000
Exposure changed by 0% to 931434,11 (930483) 11 3000 883958.850000
Exposure up to 931434 from 930483
Exposure changed by 0% to 931201,11 (931434) 11 3000 884862.300000
Exposure changed by 0% to 927966,11 (931434) 11 3000 884862.300000
Exposure changed by 0% to 930942,11 (931434) 11 2900 884862.300000
```

I'll continue testing and passing different parameters.
Thanks
BB

chisht commented:
According to your log,
```
Exposure _1253711_, 13>
Night Mode Enabled
```
It got into night mode because exposure was too high (1253711). 

Three seconds later
```
night mode start 973198, 8, 3100
```
Exposure was 973198 and color temperature was 3100.

Few seconds later,
```
Colortemp 2900
Night Mode Disabled
```
Color temperature dropped 200 below color temperature when night mode started (3100-2900). And that triggered the switch to day mode.
And then the cycle repeated.

This 200 below night mode color temperature is configurable at commandline using -B. You can try,
autonight -S -B 300 -vv

chisht commented:
@badabing2005 , you seems like a technical person.  And probably know more about cameras than I do. Are you also a C programmer? In that case, you can take a look at the source and maybe fix it. If not, maybe we can work together and come up with a universal solution. I don't mind working on it.

As for your questions.
- I'm using nightmode.sh which in turn uses nightmode on nightmode off in common_functions.sh If you think using IR LED and IR cut separately, we sure can do it as well.
- For the parameters, that is my understanding. Could very well be wrong. Take a look at "cat /proc/jz/isp/isp_info". It reports a few parameters, which are actually coming from the driver directly https://github.com/Dafang-Hacks/drivers/blob/16770a7c7f67520b85f8120adf9e0035357e7c91/isp/tx-isp/apical-isp/tx-isp-core.c#L2062 You can toggle autonight or IR cut and IR LED manually and see what changes.
- For logging in my code, I agree. Its not very useful for anyone other than myself. Well in my defense, I was doing it for myself only. I'd look into improving it. Let me know if you have any suggestion.

badabing2005 commented:
@chisht,

I finally got a chance to do some testing.
At first started testing with drawispinfo.html and realized that there are 3 settings being overlayed on WB, which made it impossible to read all the values.
I made a slight modification to the html file to break it down into 3 separate charts.

Here are my results and observations.
![autonight](https://user-images.githubusercontent.com/22603312/46181879-4b5bf200-c296-11e8-8be7-c61c65e405c6.png)

- I think we can ignore wb, wb_rg and wb_bg figures, at least for my environment they don't play any significant role.
Exposure is the main one, and then iridix.

- I started with lights on, IR LED on, and IR cut on (night mode off), I find that keeping IR LED on even in day mode, makes the transition to dark mode much smoother.
So for day mode, the best settings are:

IR LED on
IR Cut on
Grayscale off

when the lights are turned off, the exposure drops significantly, from ~920,000 ~1,270,000
And Iridix drops from 52 to 13

Once this drop is detected, it is best to first turn on the grayscale, then turn off the IR cut, always keeping IR LED on.
This also assures smooth transition, and no pink / purple images.

In night mode the exposure is around ~1,020,000 and Iridix is 5
in this mode when light is turned on, exposure is at around 880,000 and iridix jumps to 16.
To set it back to day mode, first turn on the IR Cut, and then turn off Grayscale.

Obviously these numbers reflect my environment, I'd be curious to know how other's fair.

If you can provide command line options to disable wb* and perhaps have the flexibility to set the exposure transition thresholds, light to dark and dark to light in absolute numbers as well as in percentage change, same for iridix, then everyone can use the charts to map their environments and tune the autonight.

Hope it helps.
Thanks

BB

corvy commented:
Just downloaded the new autonight bin and the config file and installed them on my Xiaofang with no light sensor. Have not tested for 24 hours yet but tried the command line just to check and I get the following: 

`[root@sofia-salong:bin]# autonight -S -vv
Night Mode Disabled
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
Unable to read isp_info
`

Anyone know why this is happening?

Shaker1978 commented:
Just wanted to say thank you! I managed to download the latest binary (from Sep. 22nd) and it works quite well. I'll have to play around with the parameters, but to be honest - I haven't understood at all, how to read the verbose output to know which parameter modifies which threshold. It's not that clear, when you look at it :)
Would you mind explaining one more time, how to read the -vv output and then assign the 7 parameters to the respective values of the output?

corvy commented:
> Just downloaded the new autonight bin and the config file and installed them on my Xiaofang with no light sensor. Have not tested for 24 hours yet but tried the command line just to check and I get the following:
> 
> > [root@sofia-salong:bin]# autonight -S -vv
> > Night Mode Disabled
> > Unable to read isp_info
> > Unable to read isp_info
> 
> Anyone know why this is happening?
> 
> FYI: the testtime is during daytime so it is supposed to not be in nightmode atm... :)

Downloaded the bin one more time and the error is gone, so now I have to wait for sun to set and we will see :)

czacha338 commented:
Guys, could you help me? 

I have downloaded the newest autonight bin from this commit and my autonight does not accept the -S function. Do you have any idea?

corvy commented:
wget with this url, that worked for me:

[https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/raw/5a43d336c796606a0d49fb54a6ddc4513bd85400/firmware_mod/bin/autonight](url)

czacha338 commented:
Thanks for help. I don't know why I could not download it before. But now I can test it. Thanks! :+1:

chisht commented:
Thanks @badabing2005  and everyone else for testing.
I was trying to make it a universal solution. That's why my logic is quite complicated. But its obvious that every lighting condition is very different. And we may not be able to find something that works for everyone.

I simplified the logic and came up with another html to help fine tuning the parameters. 

1. Please copy the firmware_mod/www/sw_night_configure.html and firmware_mod/www/cgi-bin/getispinfo.cgi to the corresponding folders on your sdcard. 
2. Disable auto night detection.
3. Visit https://x.x.x.x/sw_night_configure.html 

That page monitors Exposure, Iridix and WhiteBalance color temperature, tries to auto detect day/night and switches accordingly. It should be self explanatory. There are few parameters that can be changed. If this works, then we can easily make that html generate command-line and add those option to autonight binary.

At the bottom of the page, there also are some real-time graphs for some of the parameters.

chisht commented:
@badabing2005 
> 
> I finally got a chance to do some testing.
> At first started testing with drawispinfo.html and realized that there are 3 settings being overlayed on WB, which made it impossible to read all the values.
> I made a slight modification to the html file to break it down into 3 separate charts.
> 
> Here are my results and observations.

First of all, thanks for the detailed analysis. I changed the night_mode on/off sequence to as what you suggested. It sure looks much better.  I also removed the complicated logic and made it something like what you are suggesting. But for now, I only did that in the javascript. If that works, same logic can easily be added to the binary.

> * I think we can ignore wb, wb_rg and wb_bg figures, at least for my environment they don't play any significant role.

wb is pretty useful for me. if something is in front of the camera and night mode is on, IR reflected back is quite strong and drops the exposure significantly. Make it just look like day time. And in that case, I found that using wb is the best option.

> Obviously these numbers reflect my environment, I'd be curious to know how other's fair.

Mine is quite different. For me lights on in day mode, Iridix is low. But when night mode is enabled, Iridix jumps. Your Iridix readings are opposite of that. I'm guessing you don't have anything close to the camera that can reflect IR back.

corvy commented:
Just for follow-up, the binary and config worked great in my environment.
Night mode switched automatically just like I wanted 😎



Den søn. 30. sep. 2018, 06:31 skrev chisht <notifications@github.com>:

> @badabing2005 <https://github.com/badabing2005>
>
> I finally got a chance to do some testing.
> At first started testing with drawispinfo.html and realized that there are
> 3 settings being overlayed on WB, which made it impossible to read all the
> values.
> I made a slight modification to the html file to break it down into 3
> separate charts.
>
> Here are my results and observations.
>
> First of all, thanks for the detailed analysis. I changed the night_mode
> on/off sequence to as what you suggested. It sure looks much better. I also
> removed the complicated logic and made it something like what you are
> suggesting. But for now, I only did that in the javascript. If that works,
> same logic can easily be added to the binary.
>
>
>    - I think we can ignore wb, wb_rg and wb_bg figures, at least for my
>    environment they don't play any significant role.
>
> wb is pretty useful for me. if something is in front of the camera and
> night mode is on, IR reflected back is quite strong and drops the exposure
> significantly. Make it just look like day time. And in that case, I found
> that using wb is the best option.
>
> Obviously these numbers reflect my environment, I'd be curious to know how
> other's fair.
>
> Mine is quite different. For me lights on in day mode, Iridix is low. But
> when night mode is enabled, Iridix jumps. Your Iridix readings are opposite
> of that. I'm guessing you don't have anything close to the camera that can
> reflect IR back.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/pull/699#issuecomment-425693951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAIcNK6MBS-OHzTq8w5hBKGqG67tlLjGks5ugEkbgaJpZM4WvQZQ>
> .
>

corvy commented:
Here is my scene: https://cdn.pbrd.co/images/HGf1N3x.jpg

If interested I can capture a night screenshot as-well.

jmtatsch commented:
What a nice scene :) can you capture one night mode and one with the web ui? We could use that for the front page as marketing screenshots if you are were ok with that?

chisht commented:
@corvy , nice scene indeed. And glad to know that it works for you.

I think the logic in binary is quite complicated and the commandline parameters are even more complicated to explain. 

So I'm trying to do a simpler logic with html page that can explain parameters and help in fine tuning them for a particular use case. What I've now is the sw_night_configure.html which has all the logic to detect/change the mode. Hoping its easy to understand and can be tweak easily for any use case. 

We can go for either this simpler logic with easy tweaking or complicated logic with hard to understand parameters. Which works for most.

badabing2005 commented:
@chisht 

This is awesome, it's really nice what you've done, self explanatory and makes it very clear why you've added all the checks and complexity.
Kudos, really nice job.

The following settings are working fine for me
![good_autonight](https://user-images.githubusercontent.com/22603312/46258905-8f6e1300-c49f-11e8-820e-a56fe002d627.png)

Couple of minor comments.
I'm not sure if it is a timing issue or a sequence issue, switching from day mode to night mode, my IR cut off is applied before switching to grayscale (night) mode, roughly 1 second, 
It would be great if the sequence of those two were reversed to avoid the pink image.
If it is a timing issue, perhaps a delay param would help.

This is probably esthetic issue, but am reporting it just in case the code is interpreting it wrongly.
In night mode when exposure value drops below the threshold 1200000, it takes several readings (seconds) before the red numbers turn green

Typo: exposure not exporusre

![autonight 2](https://user-images.githubusercontent.com/22603312/46259037-3a330100-c4a1-11e8-961c-7035aed103ca.png)

Not sure if you have Home Assistant or similar setup, with MQTT, with these cameras, the light sensor value is at constant 40% most likely due to lack of LDR
perhaps night mode / day mode group of numbers can be translated to a 0-100 figure through some clever formula (exposure being highly weighted)  and that value pushed to MQTT
Obviously this has no bearing on the proper autonight operation, just being more complete.

Next step is how I turn the above figures to autonight Params.

Thanks again  for the great work.
BB

corvy commented:
I will capture any screenshot you like 😁 just use the photo as you like.
The web page is that the main page of the web UI or the new settings page?

30. sep. 2018 17:17 skrev "badabing2005" <notifications@github.com>:

@chisht <https://github.com/chisht>

This is awesome, it's really nice what you've done, self explanatory and
makes it very clear why you've added all the checks and complexity.
Kudos, really nice job.

The following settings are working fine for me
[image: good_autonight]
<https://user-images.githubusercontent.com/22603312/46258905-8f6e1300-c49f-11e8-820e-a56fe002d627.png>

Couple of minor comments.
I'm not sure if it is a timing issue or a sequence issue, switching from
day mode to night mode, my IR cut off is applied before switching to
grayscale (night) mode, roughly 1 second,
It would be great if the sequence of those two were reversed to avoid the
pink image.
If it is a timing issue, perhaps a delay param would help.

This is probably esthetic issue, but am reporting it just in case the code
is interpreting it wrongly.
In night mode when exposure value drops below the threshold 1200000, it
takes several readings (seconds) before the red numbers turn green

Typo: exposure not exporusre

[image: autonight 2]
<https://user-images.githubusercontent.com/22603312/46259037-3a330100-c4a1-11e8-961c-7035aed103ca.png>

Not sure if you have Home Assistant or similar setup, with MQTT, with these
cameras, the light sensor value is at constant 40% most likely due to lack
of LDR
perhaps night mode / day mode group of numbers can be translated to a 0-100
figure through some clever formula (exposure being highly weighted) and
that value pushed to MQTT
Obviously this has no bearing on the proper autonight operation, just being
more complete.

Next step is how I turn the above figures to autonight Params.

Thanks again for the great work.
BB

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/EliasKotlyar/Xiaomi-Dafang-Hacks/pull/699#issuecomment-425728218>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAIcNDW4VCmoZgLHHMTMzrR8_cqdZUsQks5ugOCegaJpZM4WvQZQ>
.

Shaker1978 commented:
> Thanks @badabing2005 and everyone else for testing.
> I was trying to make it a universal solution. That's why my logic is quite complicated. But its obvious that every lighting condition is very different. And we may not be able to find something that works for everyone.
> 
> I simplified the logic and came up with another html to help fine tuning the parameters.
> 
>...
> At the bottom of the page, there also are some real-time graphs for some of the parameters.

Sorry... what do I have to do, when I don't see any realtime data? My isp_info has 0 bytes... Though the autonight binary is the correct one, that accepts the -S parameter.

chisht commented:
> > Thanks @badabing2005 and everyone else for testing.
> > I was trying to make it a universal solution. That's why my logic is quite complicated. But its obvious that every lighting condition is very different. And we may not be able to find something that works for everyone.
> > I simplified the logic and came up with another html to help fine tuning the parameters.
> > ...
> > At the bottom of the page, there also are some real-time graphs for some of the parameters.
> 
> Sorry... what do I have to do, when I don't see any realtime data? My isp_info has 0 bytes... Though the autonight binary is the correct one, that accepts the -S parameter.

That's weird. Though it has nothing to do with the autonight binary. isp_info is provided by the kernel driver itself. As long as your camera is working, it should always be there.

From a shell terminal, could you run the following,
`cat /proc/jz/isp/isp_info`
If it returns with file not found, please see if you have anything else in that folder.

Shaker1978 commented:
The file is there, besides files like "cmd" "isp_de_hilight" and "isp_gamma", but all of them have 0 bytes... My camera is working, but those files have no content.

chisht commented:
> Couple of minor comments.
> I'm not sure if it is a timing issue or a sequence issue, switching from day mode to night mode, my IR cut off is applied before switching to grayscale (night) mode, roughly 1 second,
> It would be great if the sequence of those two were reversed to avoid the pink image.
> If it is a timing issue, perhaps a delay param would help.

This probably is because I'm sending those command asynchronously. What I should do is send first, wait for completion then send the second. I'll fix it hopefully next week.

> This is probably esthetic issue, but am reporting it just in case the code is interpreting it wrongly.
> In night mode when exposure value drops below the threshold 1200000, it takes several readings (seconds) before the red numbers turn green
> 

Sorry, didn't get the chance to check that. Will next week.

> Typo: exposure not exporusre

Fixed, thanks for pointing out. If I make a spelling mistake once, my editor keeps suggesting the same wrong spelling. At least it is consistent.

> Not sure if you have Home Assistant or similar setup, with MQTT, with these cameras, the light sensor value is at constant 40% most likely due to lack of LDR
> perhaps night mode / day mode group of numbers can be translated to a 0-100 figure through some clever formula (exposure being highly weighted) and that value pushed to MQTT
> Obviously this has no bearing on the proper autonight operation, just being more complete.

I am running home assistant in an LXC container. But haven't really started using it yet.
As for MQTT, we can hardly determine day or night based on what we have. The main problem is that exposure reading we get is sensitive to IR (unless we enable the IR cut filter). I'm guessing LDR is not. 
Probably best at this point would be to report just day or night for MQTT. Unless you have a clever logic in mind to convert it to light readings.


 
> Next step is how I turn the above figures to autonight Params.

Done. Please give it a try.

chisht commented:
Updated sw_night_configure.html to generate the commandline for autonight. And also updated autonight to use the same simpler logic and to accept the commandline tweaks.
Anyone willing to test it, please

- Update autonight binary from the PR
- Copy the firmware_mod/www/sw_night_configure.html and firmware_mod/www/cgi-bin/getispinfo.cgi to the corresponding folders on your sdcard.
- Disable auto night detection.
- Visit https://x.x.x.x/sw_night_configure.html It should run the autonight detection logic from your browser. Please turn the lights on/off and see if it works. If not, feel free to change the number of options on that page.
- Once you are satisfied with the working of day/night detection in the browser, please copy the "Commandline to use:" and add it to config/autonight.conf on the sdcard.
- Now you can enable the autonight and hopefully all would work.

If trying autonight from the shell, you can also add -v (or -vv) for extra verbosity.

@badabing2005 , binary is using nightmode.sh script for now. There isn't an script to turn ir_led and/or ir_cut individually. That need to be added first. Hopefully next week.

Shaker1978 commented:
Sorry, but your latest version of autonight does accept the -S parameter, but it won't start with -S -B xxxx or the generated parameter list from sw_night_configure.html (-S -j x -w x ....)

chisht commented:
> The file is there, besides files like "cmd" "isp_de_hilight" and "isp_gamma", but all of them have 0 bytes... My camera is working, but those files have no content.

Which camera is this? Maybe the firmware you have loaded is different? These files are handled by the kernel driver itself. Which is part of the firmware that you load to the camera. Nothing to do with the files you copied to the sdcard from firmware_mod folder.

badabing2005 commented:
@chisht,

Thank you, I can confirm that the new files are working as designed,
I was able to capture the command line and use it with the binary
Although it looks like I need to tune a little further because the **night mode disabled** part is unplanned.

```
autonight -S -j 3 -w 3 -1 1200000 -2 930000,14,10 -3 3000,12,4 -vv
Starting software method using
jitter_percent = 3
eq1_user_exposure = 1200000
eq2_user_exposure = 930000
eq2_user_iridix = 14
eq2_count = 10
eq3_user_wb = 3000
eq3_user_iridix = 12
eq3_count = 4
sec_wait = 3
Night Mode Disabled
(893783, 56, 3400)
jitter active
(895322, 56, 3400)
mode switch wait active
(899872, 56, 3400)
mode switch wait active
(900463, 55, 3400)
mode switch wait active
(900897, 55, 3400)
(901174, 55, 3400)
(901452, 55, 3400)
(901573, 55, 3400)
(901695, 55, 3400)
(903077, 55, 3400)
(904615, 55, 3400)
(896141, 55, 3400)
(896444, 56, 3400)
(902735, 56, 3400)
(902700, 55, 3400)
(902977, 55, 3400)
(903099, 55, 3500)
(902751, 55, 3500)
(902873, 55, 3500)
(902525, 55, 3500)
(902803, 55, 3500)
(902455, 55, 3500)
(902733, 55, 3500)
(902854, 55, 3500)
(902806, 55, 3500)
(902927, 55, 3500)
(903048, 55, 3500)
(903326, 55, 3500)
(917632, 55, 3500)
(959768, 54, 4100)
jitter active
(962432, 53, 4500)
(964095, 49, 4600)
(964588, 48, 4700)
(965247, 48, 4700)
(965572, 46, 4700)
(965692, 45, 4800)
(965642, 44, 4800)
(963829, 44, 4800)
(963825, 45, 4800)
(962669, 45, 4800)
(959234, 45, 4700)
(960909, 45, 4700)
(959424, 45, 4700)
(959749, 45, 4700)
(959581, 45, 4700)
(959906, 45, 4700)
(960231, 45, 4700)
(960062, 45, 4700)
(959071, 45, 4700)
(959396, 45, 4700)
(959721, 45, 4700)
(960046, 45, 4700)
(963736, 45, 4700)
(1179224, 30, 4700)
jitter active
(1261989, 15, 4500)
jitter active
(1262985, 13, 4400)
Switching to night new exp=1262985
Night Mode Enabled
(1022830, 5, 4200)
jitter active
(956515, 6, 4200)
jitter active
(952552, 8, 4000)
mode switch wait active
(947088, 9, 3500)
mode switch wait active
(949114, 9, 3300)
mode switch wait active
(948058, 9, 3200)
Eq2 wait_count active exp=948058 iridix=9
(944080, 10, 3100)
Eq2 wait_count active exp=944080 iridix=10
(944205, 10, 3000)
Eq2 wait_count active exp=944205 iridix=10
(944089, 10, 3000)
Eq2 wait_count active exp=944089 iridix=10
(944328, 10, 2900)
Eq2 wait_count active exp=944328 iridix=10
(944447, 10, 2900)
Eq2 wait_count active exp=944447 iridix=10
(943514, 10, 2800)
Eq2 wait_count active exp=943514 iridix=10
(941638, 10, 2800)
Eq2 wait_count active exp=941638 iridix=10
(941403, 10, 2800)
Eq2 wait_count active exp=941403 iridix=10
(942240, 10, 2800)
Eq2 wait_count active exp=942240 iridix=10
(942123, 10, 2800)
Eq2 switching to day mode exp=942123 iridix=10
Night Mode Disabled
(1262985, 8, 3200)
jitter active
(1262985, 12, 3800)
mode switch wait active
(1262985, 12, 4000)
mode switch wait active
(1262985, 12, 4200)
mode switch wait active
(1262985, 13, 4200)
Switching to night new exp=1262985
Night Mode Enabled
(1007487, 5, 4000)
jitter active
(955917, 7, 4000)
jitter active
(965223, 7, 3900)
mode switch wait active
(966065, 7, 3700)
mode switch wait active
(966077, 8, 3600)
mode switch wait active
(963376, 8, 3400)
Eq2 wait_count active exp=963376 iridix=8
(953725, 8, 3300)
Eq2 wait_count active exp=953725 iridix=8
(957991, 8, 3100)
Eq2 wait_count active exp=957991 iridix=8
(946716, 9, 3100)
Eq2 wait_count active exp=946716 iridix=9
(947071, 9, 3000)
Eq2 wait_count active exp=947071 iridix=9
(947071, 9, 3000)
Eq2 wait_count active exp=947071 iridix=9
^C
```

I think background TV with different levels of light can cause false triggers, but that's no fault of the code.
I know this is a tough ask, and probably impossible, but let me throw it out there.
Is it possible to a region / zone where the measurements are taken? Just like motion detection zone, can we eliminate the variable light (where the TV is) zone?
Probably not.

> This probably is because I'm sending those command asynchronously. What I should do is send first, wait for completion then send the second. I'll fix it hopefully next week.

Yep, I checked, and you're sending in correct order, so the issue is due to async.
I even checked `/system/sdcard/bin/setconf -k n -v 1` which is what eventually is executed to do the grayscale, and that command is very slow.
So blocking till it completes will indeed fix the issue, thanks

> Fixed, thanks for pointing out. If I make a spelling mistake once, my editor keeps suggesting the same wrong spelling. At least it is consistent.

lol

> binary is using nightmode.sh script for now. There isn't an script to turn ir_led and/or ir_cut individually. That need to be added first. Hopefully next week.

Yep, I understand that, unlike the `sw_night_configure.html` the binary still calls the nightmode.sh, which does combo of those IR led / ir cut / grayscale.

> As for MQTT, we can hardly determine day or night based on what we have. The main problem is that exposure reading we get is sensitive to IR (unless we enable the IR cut filter). I'm guessing LDR is not.
> Probably best at this point would be to report just day or night for MQTT. Unless you have a clever logic in mind to convert it to light readings.

Agreed, at this point it is best to just report day / night which is far better than 40% constant
The logic I had in mind was never accurate, was just more like what you're doing now already, if exposure is increased (value decreased), but iridix is high, then it doesn't consider it true exposure, similarly for the WB and iridix combo.
And the counters, closer we get to the threshold counter, we consider higher exposure (if we're in night mode)
Of course these number mean nothing, and it's just cosmetic.
I'll play a little and see if I can come up with anything meaningful.

@Shaker1978 
> The file is there, besides files like "cmd" "isp_de_hilight" and "isp_gamma", but all of them have 0 bytes... My camera is working, but those files have no content.

I'm sure you checked the content using the `cat` command as chisht provided the command, FYI those files being 0 bytes don't meant they don't have any content, unless you checked with cat.
Mine are 0 bytes as well, `/proc` entries are typically 0 bytes

Thanks @chisht for all the good work.
Have  a good week.
BB

Shaker1978 commented:
Thanks. It works now.
My camera is the DaFang. And the firmware I flashed is the one from this package here. cfw-1.3.bin was the filename. How can I get more details about which camera / revision I have?

chisht commented:
> Thanks. It works now.
> My camera is the DaFang. And the firmware I flashed is the one from this package here. cfw-1.3.bin was the filename. How can I get more details about which camera / revision I have?
> 
> I'm sorry. Using cat I actually get some output, thanks for the hint, badabing.
> It says:
> Software Version : H201709151519
> Firmware Version : H01-380
> SENSOR NAME : jxf22
> and some more values
> Still nothing is being displayed in your html file.

Did you copy cgi-bin/getIspInfo.cgi ?
Could you try https://x.x.x.x/cgi-bin/getIspInfo.cgi to check that file.

chisht commented:
@badabing2005 

> autonight -S -j 3 -w 3 -1 1200000 -2 930000,14,10 -3 3000,12,4 -vv
> eq2_user_exposure = 930000
> eq2_user_iridix = 14
> eq2_count = 10
> Eq2 wait_count active exp=948058 iridix=9
> (944080, 10, 3100)

Something is wrong here. How can the exposure of 944080 can trigger (exposure < eq2_user_exposure) condition. Probably a bug in my code. Will check.
Try the browser version, does it behave the same?

> I know this is a tough ask, and probably impossible, but let me throw it out there.
> Is it possible to a region / zone where the measurements are taken? Just like motion detection zone, can we eliminate the variable light (where the TV is) zone?
> Probably not.

These values are coming directly from the kernel driver for the camera sensor. As far as I know, we have no control over it. In the absence of the datasheet for the actual camera sensor module, I'm not even sure if it is supported by the hardware or not.

chisht commented:
@badabing2005
 
> > autonight -S -j 3 -w 3 -1 1200000 -2 930000,14,10 -3 3000,12,4 -vv
> > eq2_user_exposure = 930000
> > eq2_user_iridix = 14
> > eq2_count = 10
> > Eq2 wait_count active exp=948058 iridix=9
> > (944080, 10, 3100)
> 
> Something is wrong here. How can the exposure of 944080 can trigger (exposure < eq2_user_exposure) condition. Probably a bug in my code. Will check.

This indeed was a typo in the C code. Fixed.

chisht commented:
> OMG. so sorry again. I included sw_autonight... into the index.html the same way all the other subpages have been included. I added it unterneath "Motion detection", but when it is opened this way, your code doesn't work. When I type it into the browser directly (as the only page), it seems to work now. Thank you! Weird stuff...

You should try things as is first before making changes to it. I know my html sucks. Everything is so big and colors are terrible. Well, I rarely do html, have bad eye sight and couldn't care less for the colors.

> Yes, I did. I even renamed it to CamelCase after I saw your html, because it had been all lower case before. Calling the CGI directly gives me the output of the file in the browser. I don't know, if it helps, but when I open sw_autonight_config from the filesystem (no webserver) I see the running grids. When I open it from the camera, I don't even see that.
> 

From what you described, it seems like you were missing the javascript library. You should at a minimum open it in chrome/firebox debugger (right click, inspect) and look for errors.

chisht commented:
I was googling for Iridix, came across a few mentions of some dynamic range compression algorithm from a company called apical. https://www.dpreview.com/articles/3798759501/apical

Well, the driver on this camera is using libapical_isp.a. Apical is mentioned a few times in the code and we do have something called Iridix as well. Is it just a weird coincidence or is the Iridix that I thought is IR strength is actually some dynamic range algorithm. I thought apical in the code just refers to API calibration and Iridix is something like IR Index. The value we get from Iridix seems related to IR strength as well.

ZipperZ commented:
Today I was ready to go "guns blazing" debugging using the awesome web debug panel. Updated everything to the latest version. Turned on the autonight... and everything was working perfectly. @chisht amazing work, thank you!

Shaker1978 commented:
> You should try things as is first before making changes to it. I know my html sucks. Everything is so big and colors are terrible. Well, I rarely do html, have bad eye sight and couldn't care less for the colors.
You're right. But I didn't change anything, I just integrated your html into the index.html to be able to access it by clicking, instead of writing the whole URL down, which is a pain on mobiles for example.

> From what you described, it seems like you were missing the javascript library. You should at a minimum open it in chrome/firebox debugger (right click, inspect) and look for errors.
By now everything is working. There must be a problem with the existing javascript of the webUI and yours. The way, a page is called (kind of single-page-app) and the way your JS code is stored or called, I don't know. Opening your html alone works. Both, locally and on the cam.

Thanks for the good work! Never mind the big boxes and bright colors :D

akutruff commented:
Is this guy going to be merged?

jmtatsch commented:
Yes, when the source PR Is up to date at the correct repo

badabing2005 commented:
@chisht 

> This indeed was a typo in the C code. Fixed.

Yep, working really well, I think this is great already.
To make it perfect and easy for every user, here are few suggestions in my humble opinion.

> This probably is because I'm sending those command asynchronously. What I should do is send first, wait for completion then send the second. I'll fix it hopefully next week.
> @badabing2005 , binary is using nightmode.sh script for now. There isn't an script to turn ir_led and/or ir_cut individually. That need to be added first. Hopefully next week.

- Would be great if the above was added, decoupling would greatly help avoid intermediate transient states that are poor (pink screens for example)
- Would make it easy for everyone as suggested by @Shaker1978 to make it part of the main index.html, under Manage menu, this way people don't have to remember full URI to access the configuration page. (By the way, the html as is is great, don't worry about fonts and colors, it's the functionality and convenience that matters)
- Just like motion detection, once the options are saved, it would be great if they are saved into autonight.conf
- When loading the autonight configuration page, load the settings from the autonight.conf, but have an option to reset to your default values if one screws up badly and doesn't know what numbers to start with.
- When in Autonight Configuration page, **Play** should trigger Autonight Off, so that the page controls the autonight and doesn't clash with the binary autonight doing its thing.
- Similarly when **Pause** is pressed, an option to turn back the autonight (though it is preferable to not make it automatic)
- While in Pause, the numbers should continue to update, and the graphs reflect what's happening, to allow testing the binary yet see what's happening.

Of course these are just suggestions and not a demand, I'm already fully set with what you have given us.
Awesome work and thank you.

BB

chisht commented:
@badabing2005 Thanks for comments. I'd read though and try to implement what you suggested.
@jmtatsch , Will do. Sorry last couple of weeks have been pretty busy.  Do you want me to make any change or just upload my current C code to the proper repo?

jmtatsch commented:
Just upload the current version to the repo. I will compile it and check it in because not everyone should check in binaries.

chisht commented:
@jmtatsch , done. It should generate the same binary that I uploaded here. But you are right, not everyone should be allowed to check in binaries.

I was hoping to clean-up the code, take care of things that @badabing2005 pointed out and put it up in the correct repo. But haven't had the chance yet. Unless you or someone else want to do, I can get back to it hopefully in a week or so.

jmtatsch commented:
@chisht would be great if you improved software auto night over time. However, I already merged it to move things along, so i will close this pull request (sorry I didn't manage to add it to your pull request somehow).
